extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js

```diff
@@ -148,50 +150,34 @@ export class AmpAdNetworkAdsenseImpl extends AmpA4A {
   }
 
   /** @override */
-  emitLifecycleEvent(eventName, opt_associatedEventData) {
+  emitLifecycleEvent(eventName) {
+    // This function can be invoked by the superclass constructor before this
+    // class's constructor gets to run.  In that case, this.lifecycleReporter_
+    // may not be initialized.  So initialize it here.  However, we'd rather
+    // not be doing this check and initialization on every emitLifecycleEvent
+    // call, so we'll pull a classic JS move and redefine this function to
+    // the underlying sendPing once we've initialized the lifecycleReporter_.
+    // Finally, make sure that the first ping actually gets sent.
     this.lifecycleReporter_ = this.lifecycleReporter_ ||
-        this.initLifecycleReporter();
-    switch (eventName) {
-      case 'adRequestEnd':
-        const fetchResponse = opt_associatedEventData;
-        const qqid = fetchResponse.headers.get(
-            this.lifecycleReporter_.QQID_HEADER);
-        this.lifecycleReporter_.setQqid(qqid);
-        break;
-      case 'adSlotCleared':
-        this.lifecycleReporter_.sendPing(eventName);
-        this.lifecycleReporter_.reset();
-        this.element.setAttribute('data-amp-slot-index',
-            this.win.ampAdSlotIdCounter++);
-        this.lifecycleReporter_ = this.initLifecycleReporter();
-        return;
-      case 'adSlotBuilt':
-      case 'urlBuilt':
-      case 'adRequestStart':
-      case 'extractCreativeAndSignature':
-      case 'adResponseValidateStart':
-      case 'renderFriendlyStart':
-      case 'renderCrossDomainStart':
-      case 'renderFriendlyEnd':
-      case 'renderCrossDomainEnd':
-      case 'preAdThrottle':
-      case 'renderSafeFrameStart':
-        break;
-      default:
-    }
-    this.lifecycleReporter_.sendPing(eventName);
+            this.initLifecycleReporter();
+    this.emitLifecycleEvent =
+        this.lifecycleReporter_.sendPing.bind(this.lifecycleReporter_);
+    this.emitLifecycleEvent(eventName);
   }
 
+  unlayoutCallback() {
```

Comment:
Add @Override annotation

On a side note, this makes me nervous the implementors of A4A will forget to call super.unlayoutCallback which would be difficult to detect when developing/testing as it would only cause an issue should unlayoutcallback occur with a later layoutCallback execution